bigquery/google/cloud/bigquery/enums.py

```diff
@@ -0,0 +1,84 @@
+# Copyright 2019 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import re
+import warnings
+
+import enum
+import six
+
+from google.cloud.bigquery_v2.gapic import enums as gapic_enums
+
+
+def _make_sql_scalars_enum():
+    """Create an enum based on a gapic enum containing only SQL scalar types."""
+    sql_scalar_types = frozenset(
+        (
+            "INT64",
+            "BOOL",
+            "FLOAT64",
+            "STRING",
+            "BYTES",
+            "TIMESTAMP",
+            "DATE",
+            "TIME",
+            "DATETIME",
+            "NUMERIC",
+        )
+    )
+    excluded_members = frozenset(
+        ("TYPE_KIND_UNSPECIFIED", "GEOGRAPHY", "ARRAY", "STRUCT")
+    )
+
+    # Sanity check - we do not want the new enum to go out of sync with the original
+    # enum from gapic.
+    # ASSUMPTION: No existing types are ever renamed or deleted, we only try to
+    # detect cases when new types are introduced.
+    gapic_names = set(m.name for m in gapic_enums.StandardSqlDataType.TypeKind)
+    anticipated_names = sql_scalar_types | excluded_members
+    unhandled_names = gapic_names - anticipated_names
+
+    if unhandled_names:
+        msg = (
+            "The StandardSqlDataTypes enum migh be out of sync with the "
+            "original StandardSqlDataType.TypeKind enum from gapic. Check "
+            "enum members: {}".format(", ".join(unhandled_names))
+        )
+        warnings.warn(msg, UserWarning)
```

Comment:
There's not much a user can do about this error, right? It's more that we'll need to notice this error when the autogen PRs from the yoshi bot come in.

What if we did this check as a unit test?

* `sql_scalar_types` is a module-level constant like `_SQL_SCALAR_TYPES`.
* `excluded_members` is also a module-level constant like `_SQL_NONSCALAR_TYPES`.
* The test:

```
gapic_names = set(
    type_.name for type_ in gapic_enums.StandardSqlDataType.TypeKind
)
anticipated_names = (
    module_under_test._SQL_SCALAR_TYPES
    | module_under_test._SQL_NONSCALAR_TYPES
)
unhandled_names = gapic_names - anticipated_names
assert not unhandled_names
```

Reply:
Indeed, the warning was meant for us, not the users. But then again, a loud test failure is better suited for this, will refactor. :+1: